src/prepare.ts

```diff
@@ -68,55 +100,85 @@ async function prepareCompositeElement({ type, props }, errorHandler, context) {
       }
     }
   }
-  const instance = createCompositeElementInstance({ type, props }, context);
+  const instance = createCompositeElementInstance(element, context);
   return [...renderCompositeElementInstance(instance, context), preparePromise];
 }
 
-async function prepareElement(element, errorHandler, context) {
+async function prepareElement(
+  element: ReactNode,
+  errorHandler: (error: unknown) => void,
+  context: PrepareContext,
+): Promise<[ReactNode, PrepareContext, Promise<void>?]> {
   switch (getElementType(element)) {
     case ELEMENT_TYPE.NOTHING:
     case ELEMENT_TYPE.TEXT_NODE: {
       return [null, context];
     }
     case ELEMENT_TYPE.DOM_ELEMENT:
     case ELEMENT_TYPE.FRAGMENT: {
-      return [element.props.children, context];
+      return [(element as ReactElement).props.children, context];
     }
     case ELEMENT_TYPE.CONTEXT_PROVIDER: {
-      const _providers = new Map(context._providers);
-      _providers.set(element.type._context.Provider, element.props);
-      return [element.props.children, { _providers }];
+      const providerElement = element as ProviderElement;
+      const _providers: ContextProviderMap = new Map(context._providers);
+      _providers.set(
+        providerElement.type._context.Provider,
+        providerElement.props,
+      );
+      return [providerElement.props.children, { _providers }];
     }
     case ELEMENT_TYPE.CONTEXT_CONSUMER: {
-      const value = getContextValue(context, element.type._context);
+      const consumerElement = element as ConsumerElement;
+      const value = getContextValue(context, consumerElement.type._context);
 
-      const consumerFunc = element.props.children;
+      const consumerFunc = consumerElement.props.children;
       return [consumerFunc(value), context];
     }
     case ELEMENT_TYPE.FORWARD_REF: {
-      return [element.type.render(element.props, element.ref), context];
+      const forwardRefElement = element as ForwardRefElement;
+      return [
```

Comment:
You could probably get rid of a lot of this casting if you use type guards in `getElementType` like with `ELEMENT_TYPE.TEXT_NODE`

Reply:
True, type guards functions could be used, but it doesn't seem like typescript can infer the types when using a switch like this, so it would require a bit of refactoring. 

I might fix it later, but it's well tested and works now, so it's not really important.